src/python/ksc/cgen.py

```diff
@@ -255,4 +256,101 @@ def join_args(sep, callable):
 
 
 def generate_cpp_vmap_entry_point(cpp_function_name, decl):
-    raise NotImplementedError("vmap")
+    def add_vmap_dimension(t: Type):
+        if t.is_scalar:
+            return Type.Tensor(1, t)
+        if t.is_tensor:
+            return Type.Tensor(t.tensor_rank + 1, t.tensor_elem_type)
+
+        # Not clear how to to the remainder -- let's see a use case before deciding
+        raise ValueError("Vmap understands only tensors for now")
+
+    in_arg_types = arg_types_of_decl(decl)
+
+    # Add a "vmap dimension" to each arg
+    arg_types = tuple(add_vmap_dimension(a) for a in in_arg_types)
+    ks_types = tuple(ks_cpp_type(a) for a in arg_types)
+
+    num_args = len(arg_types)
+
+    def join_args(callable, sep=", "):
+        return sep.join(callable(k) for k in range(num_args))
+
+    def concat_args(callable):
+        return "".join(callable(k) for k in range(num_args))
+
+    ks_name = utils.encode_name(decl.name.mangled())
+
+    # torch::Tensor entry_my_kernel(torch::Tensor arg0, ..., torch::Tensor arg7)
+    cpp_function = f"""
+torch::Tensor {cpp_function_name}({join_args(lambda k: f'torch::Tensor arg{k}')})
+    """
+
+    cpp_declaration = f"{cpp_function};\n"
+
+    cpp = f"""
+{cpp_function} {{
+    int64_t n = arg0.size(0);
+"""
+
+    # auto* arg_data0 = arg0.data_ptr<float>();
+    # ...
+    # auto* arg_data7 = arg7.data_ptr<float>();
+    for k in range(num_args):
+        cpp += f"""
+    KS_ASSERT(arg{k}.is_contiguous());
+    KS_ASSERT(arg{k}.scalar_type() == scalar_type_of_Float);
+    KS_ASSERT(arg{k}.size(0) == n);
+
+    auto ks_arg{k} = convert_argument<{ks_types[k]}>(arg{k});
+"""
+
+    # Difficulty: depending on the rank of ks_ret, ks_ret[i] returns either
+    #   Rank 1: a reference to the float at  index i
+    #   Rank 2: an rvalue representing a view on subtensor i.
+    # inplace add must act differently on each, so we switch on the return dimension here
+
+    ks_return_type = add_vmap_dimension(decl.return_type)
+    ks_return_dim = ks_return_type.tensor_rank
+    if ks_return_dim == 1:
+        cpp += f"""
+    // Create Torch return value
+    auto ret = torch::zeros({{n}});
+    ks::Float* ret_ptr = ret.data_ptr<ks::Float>();
+
+    for (int i = 0; i != n; ++i)
+        ret_ptr[i] = ks::{ks_name}(&g_alloc {concat_args(lambda k: f", ks_arg{k}[i]")});
+
+    return ret;
+}}
+"""
+    else:
+        ks_sizes = ", ".join([f"size{d}" for d in range(ks_return_dim - 1)])
+        cpp += f"""
+    KS_ASSERT(n > 0); // TODO: Zero-size tensors
+
+    // Make the first call to determine output size
+    auto ret0 = ks::{ks_name}(&g_alloc {concat_args(lambda k: f", ks_arg{k}[0]")});
+
+    // Create Torch return value
+    auto [{ks_sizes}] = ret0.size();
+    auto ret = torch::zeros({{n, {ks_sizes}}});
+    // And wrap it in ks - this is a view of the torch data, so convert_argument, not convert_return_value
+    auto ks_ret = convert_argument<ks::tensor<{ks_return_dim}, Float>>(ret);
```

Comment:
I'm starting to think I chose poor names for the functions `convert_argument` and `convert_return_value`, as the names reflect what the functions were originally used for, not what they actually do. Maybe rename to `convert_to_ks` and `convert_from_ks`?